pmlc/dialect/tile/transforms/algebraic_opt.cc

```diff
@@ -0,0 +1,79 @@
+// Copyright 2020, Intel Corporation
+
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Support/DebugStringHelper.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "pmlc/dialect/tile/ir/ops.h"
+#include "pmlc/dialect/tile/transforms/pass_detail.h"
+#include "pmlc/dialect/tile/transforms/passes.h"
+#include "pmlc/util/logging.h"
+
+using namespace mlir; // NOLINT
+
+namespace pmlc::dialect::tile {
+
+namespace {
+
+// From:
+// %zero = tile.constant(0.0 : f64)
+// %0 = tile.contract add, mul, %zero, %I, %K
+// %1 = tile.add %0, %B
+// %2 = tile.relu %1
+// Into:
+// %0 = tile.contract add, mul, %B, %I, %K
+// %2 = tile.relu %0
+struct AddInitPattern final : public OpRewritePattern<AddOp> {
+  using OpRewritePattern<AddOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(AddOp op, PatternRewriter &rewriter) const {
+    if (ContractionOp contractOp =
+            dyn_cast_or_null<ContractionOp>(op.lhs().getDefiningOp())) {
+      if (!contractOp->hasOneUse())
+        return failure();
+      if (isa_and_nonnull<ContractionOp>(op.rhs().getDefiningOp()))
```

Comment:
I'm not sure I understand why RHS can't be a contraction.

Reply:
There's probably a better way to handle this case, but the case I was running into was `concat`. 

```
%0 = tile.contract %zero, ...
%1 = tile.contract %zero, ...
%2 = tile.add %0, %1
```

The result would be clearly broken:
```
%0 = tile.contract, %1, ...
%1 = tile.contract, %zero, ...
```